src/util/thpool/thpool.c

```diff
@@ -41,207 +42,222 @@ static volatile int threads_on_hold;
-  if (thpool_p == NULL) return;
-
-  volatile int threads_total = thpool_p->num_threads_alive;
-
-  /* End each thread 's infinite loop */
-  threads_keepalive = 0;
-
-  /* Give one second to kill idle threads */
-  double TIMEOUT = 1.0;
-  time_t start, end;
-  double tpassed = 0.0;
-  time(&start);
-  while (tpassed < TIMEOUT && thpool_p->num_threads_alive) {
-    bsem_post_all(thpool_p->jobqueue.has_jobs);
-    time(&end);
-    tpassed = difftime(end, start);
-  }
-
-  /* Poll remaining threads */
-  while (thpool_p->num_threads_alive) {
-    bsem_post_all(thpool_p->jobqueue.has_jobs);
-    sleep(1);
-  }
-
-  /* Job queue cleanup */
-  jobqueue_destroy(&thpool_p->jobqueue);
-  /* Deallocs */
-  int n;
-  for (n = 0; n < threads_total; n++) {
-    thread_destroy(thpool_p->threads[n]);
-  }
-  free(thpool_p->threads);
-  free(thpool_p);
+	/* No need to destory if it's NULL */
+	if(thpool_p == NULL) return;
+
+	volatile int threads_total = thpool_p->num_threads_alive;
+
+	/* End each thread 's infinite loop */
+	threads_keepalive = 0;
+
+	/* Give one second to kill idle threads */
+	double TIMEOUT = 1.0;
+	time_t start, end;
+	double tpassed = 0.0;
+	time(&start);
+	while(tpassed < TIMEOUT && thpool_p->num_threads_alive) {
+		bsem_post_all(thpool_p->jobqueue.has_jobs);
+		time(&end);
+		tpassed = difftime(end, start);
+	}
+
+	/* Poll remaining threads */
+	while(thpool_p->num_threads_alive) {
+		bsem_post_all(thpool_p->jobqueue.has_jobs);
+		sleep(1);
+	}
+
+	/* Job queue cleanup */
+	jobqueue_destroy(&thpool_p->jobqueue);
+	/* Deallocs */
+	int n;
+	for(n = 0; n < threads_total; n++) {
+		thread_destroy(thpool_p->threads[n]);
+	}
+	free(thpool_p->threads);
+	free(thpool_p);
 }
 
 /* Pause all threads in threadpool */
 void thpool_pause(thpool_* thpool_p) {
-  int n;
-  for (n = 0; n < thpool_p->num_threads_alive; n++) {
-    pthread_kill(thpool_p->threads[n]->pthread, SIGUSR2);
-  }
+	int n;
+	for(n = 0; n < thpool_p->num_threads_alive; n++) {
+		pthread_kill(thpool_p->threads[n]->pthread, SIGUSR2);
+	}
 }
 
 /* Resume all threads in threadpool */
 void thpool_resume(thpool_* thpool_p) {
-  // resuming a single threadpool hasn't been
-  // implemented yet, meanwhile this supresses
-  // the warnings
-  (void)thpool_p;
+	// resuming a single threadpool hasn't been
+	// implemented yet, meanwhile this supresses
+	// the warnings
+	(void)thpool_p;
 
-  threads_on_hold = 0;
+	threads_on_hold = 0;
 }
 
 int thpool_num_threads_working(thpool_* thpool_p) {
-  return thpool_p->num_threads_working;
+	return thpool_p->num_threads_working;
+}
+
+int thpool_num_threads(thpool_* thpool_p) {
+	return thpool_p->num_threads_alive;
+}
+
+int thpool_get_thread_id(thpool_* thpool_p, pthread_t pthread) {
+	for(int i = 0; i < thpool_p->num_threads_alive; i++) {
+		thread *thread = thpool_p->threads[i];
+		if(thread->pthread == pthread) return thread->id;
+	}
+
+	err("thpool_get_thread_id(): Could not locate thread\n");
+	assert(0);
```

Comment:
If the query was issued in a MULTI block or Lua script, we can validly get to this point.

Reply:
👍